docs/source/design/threaded_runner.rst

```diff
@@ -56,25 +56,41 @@ The main thread waits on the ``output_queue`` for data or signals and handles th
 Protocols
 =========
 
-Due to its history the runner implementation uses the interface defined in ``SubprocessProtocol`` (asyncio.protocols.SubprocessProtocol) (although the sub process protocol interface is defined in the asyncio libraries, the current thread-runner implementation does not make use of ``async``).
+Due to its history datalad uses the protocol defined in ``asyncio.protocols.SubprocessProtocol`` and in ``asyncio.protocols.BaseProtocol``.
+To keep compatibility with the code base, the threaded-runner implementation uses
+the same interface. Please note, although we use the same interface and
+although the interface is defined in the asyncio
+libraries, the threaded-runner implementation does not make any use of ``asyncio``.
+The description of the interface nevertheless applies in the context of the threaded-runner.
+The following methods of the ``SubprocessProtocol`` are supported.
 
     - ``SubprocessProtocol.pipe_data_received(fd, data)``
     - ``SubprocessProtocol.pipe_connection_lost(fd, exc)``
     - ``SubprocessProtocol.process_exited()``
 
-In addition the methods of ``BaseProtocol`` are called, i.e.:
+In addition the following methods of ``BaseProtocol`` are supported:
 
     - ``BaseProtocol.connection_made(transport)``
     - ``BaseProtocol.connection_lost(exc)``
 
 
-The datalad-provided protocol ``WitlessProtocol`` provides an additional callback:
+The datalad-provided protocol ``datalad.runners.protocol.WitlessProtocol``
+provides an additional callback:
 
     - ``WitlessProtocol.timeout(fd)``
 
-The method ``timeout()`` will be called when the parameter ``timeout`` in ``WitlessRunner.run``, ``ThreadedRunner.run``, or ``run_command`` is set to a number specifying the desired timeout in seconds. If no data is received from ``stdin``, or ``stderr`` (if those are supposed to be captured), the method ``WitlessProtocol.timeout(fd)`` is called with ``fd`` set to the respective file number, e.g. 1, or 2. If ``WitlessProtocol.timeout(fd)`` returns ``True``, the file descriptor will be closed and the associated threads will exit.
+The method ``timeout()`` will be called when the parameter ``timeout`` in
+``WitlessRunner.run``, ``ThreadedRunner.run``, or ``run_command`` is set to a
+number specifying the desired timeout in seconds. If no data is received from
+``stdin``, or ``stderr`` (if those are supposed to be captured), the method
+``WitlessProtocol.timeout(fd)`` is called with ``fd`` set to the respective
+file number, e.g. 1, or 2. If ``WitlessProtocol.timeout(fd)`` returns ``True``,
+the file descriptor will be closed and the associated threads will exit.
```

Comment:
```suggestion
only the corresponding file descriptor will be closed and the associated threads will exit.
```

so it is responsibility of what code to close the other fd?  e.g. if timeout happens on fd 2 but not fd 1, whenever both are captured, what would happen?

Reply:
> if timeout happens on fd 2 but not fd 1, whenever both are captured, what would happen?

fd 1 would be left untouched. It will be closed latest if the runner exits. But it might be closed earlier by the user or if a timeout appears on fd 1 and the timeout handler returns `True` for fd 1